### PR TITLE
Configure standard Prometheus scraping

### PR DESCRIPTION
**Pull request by howardjohn**
API: https://github.com/istio/api/pull/1354
Design: https://docs.google.com/document/d/1TTeN4MFmh4aUYYciR4oDBTtJsxl5-T5Tu3m3mGEdSo8/edit
Issue: https://github.com/istio/istio/issues/21366

This change modifies the Istio configurations to use "standard" `prometheus.io` annotations for scraping. This most involves adding this annotation to all installed deployments (Istiod/gateway/mixer/sidecars).

However, we cannot just add the annotation to the sidecar as a user may already have one existing. Instead, we will store their existing annotations, and during scraping istio-agent will merge both Envoy and their application metrics (as configured by `prometheus.io` annotation, if present).

The end result of this change is that prometheus scraping will work out of the box for the majority of prometheus installations (specifically, those that are configured to scrape on `prometheus.io` annotations).

A major implication of this is that scraping of application metrics will no longer be using TLS. Envoy metrics are also plain text, but they have been prior to this change. Additionally, if the application itself exposes Envoy metrics, they will conflict with istio Envoy metrics - this typically only happens when wrapping another Envoy ingress like Ambassador or similar. For these two cases (or any other reason), there are two ways to opt out:
* mesh wide, with `--set values.meshConfig.enablePrometheusMerge=false`
* As an annotation on the pod, with `prometheus.istio.io/merge-metrics: "false"`

If you do opt out, please file an issue with an explanation of your use case so we can better understand how users are using prometheus and if we can make some improvements.

### CODE REVIEW

**Comment by howardjohn** [comment]
Testing is added in previous PRs: #22268 and #22331

File: go.mod
Comment (howardjohn) [inline go.mod]: This will be reverted once the API change lands

File: manifests/istio-telemetry/prometheus/templates/configmap.yaml
Comment (howardjohn) [inline manifests/istio-telemetry/prometheus/templates/configmap.yaml]: I changed the scraping config to match what is used in stable/prometheus. This is largely the same as what we had before with the removal of the secure scraping job and all of the Istio specific stuff.

File: manifests/istio-telemetry/prometheus/templates/deployment.yaml
Comment (howardjohn) [inline manifests/istio-telemetry/prometheus/templates/deployment.yaml]: This is not needed anymore, as we bypass the application tls

File: tests/integration/telemetry/stats/prometheus/http/stats.go
Comment (howardjohn) [inline tests/integration/telemetry/stats/prometheus/http/stats.go]: This comment no longer applies, so we no longer need to set this up

File: tests/integration/security/cert_provision_prometheus/cert_provision_prometheus_test.go
Comment (howardjohn) [inline tests/integration/security/cert_provision_prometheus/cert_provision_prometheus_test.go]: This test is no longer valid, as prometheus doesn't need a cert

**Comment by howardjohn** [comment]
@mandarjog @kyessenov @douglas-reid this PR is ready to review. I have it in WIP now only because its pulling in howardjohn/api - once https://github.com/istio/api/pull/1354 merges this is ready to go

**Review by douglas-reid — COMMENTED** [review]
I know you covered this in a few of the linked artifacts, but can you also update the PR description to capture the following:

1. The implication for users, including:
    - Istio will merge envoy prometheus metrics with application metrics, exposing a single scrape target
    - The new scrape target will be clear-text (which is why secure scraping is removed).
    - How to opt-in/out.
1. A statement on the overall direction of removing Istio-specific configuration of Prometheus.

I think that would be a big help for people that don't want to click around and digest docs / comments / etc. I suspect this PR will be referenced later in release notes, etc.

**Comment by howardjohn** [comment]
@douglas-reid good idea. I have updated the description. Much of this will also need to make it into istio.io as well

File: manifests/istio-telemetry/prometheus/templates/configmap.yaml
Comment (douglas-reid) [inline manifests/istio-telemetry/prometheus/templates/configmap.yaml:43]: this is one thing we are losing here -- the dropping of labels on `envoy-stats` collection. this essentially, as best I can tell, reverts the PR that did that for 1.5.

This will result in `istio_` metrics having duplications like `{"app"="foo", "destination_app"="foo"}` etc., plus that pod template hash label, right?  And I think also add in some of the k8s labels like `kubernetes_namespace` and `pod_name`, right?

That will increase storage requirements and the overhead of bytes transferred, as well as some potential user confusion.

@mandarjog thoughts here?

Comment (howardjohn) [reply manifests/istio-telemetry/prometheus/templates/configmap.yaml:43]: The easy fix without context of other (possible) changes that we are making is just to drop these labels in the config again here. This will mean that our prom config is "better" than out-of-the-box configs though, which would be especially bad if we move forward with the addon changes.

I think its somewhat reasonable to say that if you are a user that cares about the metrics size a lot, you can configure it to drop metrics. The unfortunate thing here is the confusion (app vs destination_app) which impacts new users as well

File: pkg/kube/inject/webhook.go
Comment (douglas-reid) [inline pkg/kube/inject/webhook.go:630]: nit: do we want to monitor these failures with a metric for the injector ?

File: pilot/cmd/pilot-agent/status/server.go
Comment (howardjohn) [reply pilot/cmd/pilot-agent/status/server.go:232]: Yes, its one of the goals of this, but wasn't mentioned too much. We used to have SDS metrics, and there are certainly metrics we could add here (meta-mertics) and other areas in agent. The intent was to make this easy to add in the future but not do it in this PR

File: pilot/cmd/pilot-agent/status/server.go
Comment (howardjohn) [reply pilot/cmd/pilot-agent/status/server.go:241]: As far as I can tell prometheus is not sending auth headers? Or did I miss some? I was looking at https://github.com/prometheus/prometheus/blob/master/scrape/scrape.go#L545 and roughly following this

Comment (douglas-reid) [reply pilot/cmd/pilot-agent/status/server.go:241]: The scrape config supports Auth configuration. How that gets translated into their code is anyone's guess:

> Sets the `Authorization` header on every scrape request with the configured username and password.
...
basic_auth:
...
Sets the `Authorization` header on every scrape request with the configured bearer token. It is mutually exclusive with `bearer_token_file`.
[ bearer_token: <secret> ]
...
Sets the `Authorization` header on every scrape request with the bearer token read from the configured file. It is mutually exclusive with `bearer_token`.
[ bearer_token_file: /path/to/bearer/token/file ]

-- https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config

Comment (howardjohn) [reply pilot/cmd/pilot-agent/status/server.go:241]: Interesting, thanks. it seems like if you are creating a scrape config based on the prometheus.io annotations, it ought to be generic. It seems unlikely that you would have every app in the entire cluster responding to the same Auth headers? This is more of a guess though, unfortunately don't have a lot of data here. I would guess that if people did this, they would be customizing this a bit to add some other indicator like `prometheus.io/use-basic-auth` or something like that as well, but again just a guess. 

The challenging thing here is that users can configure prom however they want. We cannot reasonably support all of these out of the box, but the most common ones we should. I guess the hard part is identifying those cases.

I believe @Stono does something like this with different annotations, your feedback here would be appreciated

Comment (douglas-reid) [reply pilot/cmd/pilot-agent/status/server.go:241]: Fair enough. I'm definitely not sure how widespread authn is the Prom world either, let alone in combination with annotation-based SD.  I think some impls use it just to provide a low bar for protection across the fleet, but that's just a rough guess based on some quick Googling.

I was just wondering if forwarding the `Authentication` header might be worthwhile, as I don't think it is dangerous and could potentially ease some use cases. But we don't need to block on this question.

Comment (howardjohn) [reply pilot/cmd/pilot-agent/status/server.go:241]: Yeah I almost want to just forward all headers? I'll take a closer look at the prom implementation and see what other headers are possible

File: pkg/kube/inject/webhook.go
Comment (howardjohn) [reply pkg/kube/inject/webhook.go:630]: turns out we already have a `validateAnnotations` function. Let me add it there. In this function we are failing hard rather than silently containing which seems reasonable. maybe we can plug this into `istioctl analyze` as well

File: pilot/cmd/pilot-agent/status/server.go
Comment (howardjohn) [reply pilot/cmd/pilot-agent/status/server.go:232]: I created an issue + TODO in the code: https://github.com/istio/istio/issues/22825

**Comment by howardjohn** [comment]
@douglas-reid I updated the PR with some minor tweaks you suggested and some additional docs. The two open questions that remain are what to do about auth/headers in general, and the label dropping. Opinions from @istio/wg-policies-and-telemetry-maintainers would be great

File: pkg/kube/inject/webhook.go
Comment (howardjohn) [reply pkg/kube/inject/webhook.go:630]: analyzer: https://github.com/istio/istio/pull/22826

**Comment by douglas-reid** [comment]
@mandarjog @nrjpoddar @Stono what do you think re: label dropping? Can we revert to not drop the pod labels from the endpoints in the default scrape config?

File: manifests/istio-telemetry/prometheus/templates/configmap.yaml
Comment (mandarjog) [inline manifests/istio-telemetry/prometheus/templates/configmap.yaml:45]: Nit: keep job name at the top

**Comment by mandarjog** [comment]
The goal is to use external prom then we need to supply “additional scrape config”, and federation config in the future in several supported forms, one config when prom operator is used , one with something else. @howardjohn / this PR does not have to do any of that. We should take care of it later.

Given that 
1. Why do we have previous scrape config at all?
2. If we are keeping it, why reduce it in some way?

File: manifests/istio-telemetry/prometheus/templates/configmap.yaml
Comment (nrjpoddar) [inline manifests/istio-telemetry/prometheus/templates/configmap.yaml:38]: Are these recommended upstream Prometheus settings?

**Review by nrjpoddar — COMMENTED** [review]
> @mandarjog @nrjpoddar @Stono what do you think re: label dropping? Can we revert to not drop the pod labels from the endpoints in the default scrape config?

I think migration would be an issue for existing customers if they want to preserve the label set they have with 1.5 or 1.4. But tbh most of the customers are probably tinkering with the Prometheus config and not just using what Istio "demo" gives them, so their old settings will be preserved as long as they do the merge. For our customers, we provide our own version of Prometheus with custom config so this doesn't impact us.

The bigger impact is probably on UX and confusion due to similarly named labels, may be we can add a doc with link to the scrape config to fix it?

File: manifests/istio-telemetry/prometheus/templates/configmap.yaml
Comment (howardjohn) [reply manifests/istio-telemetry/prometheus/templates/configmap.yaml:43]: @nrjpoddar responding here to 

> I think migration would be an issue for existing customers if they want to preserve the label set they have with 1.5 or 1.4. But tbh most of the customers are probably tinkering with the Prometheus config and not just using what Istio "demo" gives them, so their old settings will be preserved as long as they do the merge. For our customers, we provide our own version of Prometheus with custom config so this doesn't impact us.

I would agree I don't think the "production user" is impacted here. People who care about the scalability concerns of additional labels will have their own configs. So to me this is less important.

> The bigger impact is probably on UX and confusion due to similarly named labels, may be we can add a doc with link to the scrape config to fix it?

This I would agree. I am not sure how much of an impact this is for users to decide what we should do though.

it would be easy to add the label dropping back in theory but the problem is we then would either drop application metric labels (not desirable) or have a custom scrape config for envoy (what this entire PR is trying to avoid), so its pretty challenging

File: manifests/istio-telemetry/prometheus/templates/configmap.yaml
Comment (howardjohn) [reply manifests/istio-telemetry/prometheus/templates/configmap.yaml:38]: This is copied form stable/prometheus. Its identical to ours except they have `__meta_kubernetes_namespace==default` check. I do not think there is any impact (positive or negative) of this?

**Comment by howardjohn** [comment]
> may be we can add a doc with link to the scrape config to fix it

I think we can extend https://preliminary.istio.io/docs/ops/best-practices/observability/#using-prometheus-for-production-scale-monitoring to include info about how to drop labels? what do you think?

**Comment by howardjohn** [comment]
@mandarjog I may be misunderstand this a bit but:

> Why do we have previous scrape config at all?

Our demo deployment needs some scraping configuration, or it will scrape nothing and not work.

> If we are keeping it, why reduce it in some way?

The motivation for changing our config was to align with what is in helm and other standard scraping configurations so that we can improve testing of these types of configs and make the demo experience more similar to what people will run with production setups.


> then we need to supply “additional scrape config”, and federation config in the future in several supported forms

The whole point is there is no "additional scrape config"? There is just the standard `prometheus.io/...` scraping config which is ubiquitous. We will also document how that is set up for people that are not using standard setups and for general understanding. Federation is already documented by https://preliminary.istio.io/docs/ops/best-practices/observability/#using-prometheus-for-production-scale-monitoring?

**Comment by douglas-reid** [comment]
> > may be we can add a doc with link to the scrape config to fix it
> 
> I think we can extend https://preliminary.istio.io/docs/ops/best-practices/observability/#using-prometheus-for-production-scale-monitoring to include info about how to drop labels? what do you think?

I think that document will definitely need to be updated in light of the general set of changes around add-ons and I think that is a good place for this kind of information

**Comment by douglas-reid** [comment]
FWIW, dropping pod labels had a negative impact on Kiali (https://github.com/istio/istio/issues/22584), so restoring them would whack that mole at least.

File: manifests/istio-telemetry/prometheus/templates/configmap.yaml
Comment (nrjpoddar) [reply manifests/istio-telemetry/prometheus/templates/configmap.yaml:38]: It looks equivalent.

**Comment by howardjohn** [comment]
@douglas-reid @mandarjog I am waiting for review here. If there is something you are waiting on me to change please let me know

**Comment by douglas-reid** [comment]
@howardjohn I was waiting for confirmation that everyone else was OK with the changes. I think @nrjpoddar is, but I want to be sure @mandarjog is on-board as well.

**Comment by howardjohn** [comment]
Per discussion with @mandarjog I have made this disabled by default. It can be enabled with `--set values.meshConfig.enablePrometheusMerge=true`. The branch with it enabled by default can be found in howardjohn:prometheus/standard-annotations-default for future reference when we want to enable it by default.

**Comment by howardjohn** [comment]
@mandarjog this is ready now with off by default

**Review by douglas-reid — APPROVED** [review]
I believe my concerns have been addressed.

**Comment by nrjpoddar** [comment]
Approved as long as there are docs changes as requested above. I don't want to overlook/miss doc edits. Please attach the docs PR if you can.

Not sure what's the policy but I would like to hold this PR until the docs PR is opened.

**Comment by howardjohn** [comment]
> Not sure what's the policy but I would like to hold this PR until the docs PR is opened.

SGTM, I will start working on this. I agree we should probably do this for more PRs, especially impactful ones like this. Although we should not wait for the docs to be approved, at least in the interest of code freeze

**Comment by nrjpoddar** [comment]
> > Not sure what's the policy but I would like to hold this PR until the docs PR is opened.
> 
> SGTM, I will start working on this. I agree we should probably do this for more PRs, especially impactful ones like this. Although we should not wait for the docs to be approved, at least in the interest of code freeze

Agree just an existence of istio.io PR is good by me. We need the docs PR for dev community testing too.

**Review by costinm — APPROVED** [review]

**Comment by howardjohn** [comment]
@nrjpoddar @douglas-reid docs PR: https://github.com/istio/istio.io/pull/7112

**Review by nrjpoddar — APPROVED** [review]

File: manifests/charts/global.yaml
Comment (nmittler) [inline manifests/charts/global.yaml:567]: nit: carriage return

File: pilot/cmd/pilot-agent/status/server.go
Comment (nmittler) [inline pilot/cmd/pilot-agent/status/server.go:128]: nit: `it's`

**Review by nmittler — APPROVED** [review]

**Comment by desaintmartin** [comment]
I am curious, will this become the default or recommended installation way for new installations at some point?

**Comment by howardjohn** [comment]
@desaintmartin the intent is that it is default in 1.7

**Comment by desaintmartin** [comment]
Thanks! I've seen there are discussions in the related issue about those labels and PrometheusOperator, maybe we should continue there, but what are the implication for people wanting to manage their Prometheus using a Prometheus Operator (either outside of the Istio installation or using `prometheusOperator.enable: true`)? Should they configure their Prometheus to read those annotations?

File: pilot/cmd/pilot-agent/status/server.go
Comment (cagataygurturk) [inline pilot/cmd/pilot-agent/status/server.go:249]: @howardjohn I am not sure if it is good idea to return 500 for the metrics endpoint when application metrics can't be fetched. Without metrics merge enabled, if the application has issues to serve HTTP requests, Envoy will keep delivering metrics and users are able to detect any issue when Envoy reports failing requests. With this design, when the application stops responding to requests for some reason, Prometheus will fail to scrape the whole Pod together with Envoy metrics and users will never see 5xx requests coming from this Pod.

Comment (howardjohn) [reply pilot/cmd/pilot-agent/status/server.go:249]: @cagataygurturk that is a great point. One concern I have if we switch the behavior here is that we may obscure the fact that the application is failing to respond. I have confirmed prometheus will no include the metrics returned in the case of non-200 error codes: https://github.com/prometheus/prometheus/blob/b30654211cf99d244775c25f353d4becdb1aec99/scrape/scrape.go#L560. Would love your feedback here

Comment (cagataygurturk) [reply pilot/cmd/pilot-agent/status/server.go:249]: Before writing my initial comment I also thought about your concern. In my opinion, as a user we observe the whole Pod, not specifically the app container. Even though the app container fails to respond to metrics requests, there are still Envoy metrics I would like to see there and in fact thanks to those Envoy metrics I will be able to detect that app container has a serious issue. 

When the whole metric endpoint disappears just because app container's metric endpoint fails, there is a high chance that users are not alerted because the whole metrics endpoint also disappears from Prometheus when a Pod is gracefully replaced during deployments etc. People do not tend to set up alerts for this. In this case we have a risk of keeping a failing Pod in the pool and we will never know that we are sending 5xx's.

In my opinion if app container stops reporting metrics, Envoy metrics should continue to be reported. Users can still use `scrape_samples_scraped` metric to set up alerting policies if they are interested in events where application metrics from a Pod suddenly disappears.

Comment (howardjohn) [reply pilot/cmd/pilot-agent/status/server.go:249]: We can also add metrics about the scraping with https://github.com/istio/istio/pull/24798. So I think it makes a lot of sense to continue when there are errors. Any opinions @douglas-reid?

Comment (howardjohn) [reply pilot/cmd/pilot-agent/status/server.go:249]: Thanks for the feedback. Can you take a look at https://github.com/istio/istio/pull/25668?

**Pull request merged**
